src/HexagonAlignLoads.cpp

```diff
@@ -0,0 +1,242 @@
+            int rem_mod = mod_imp(mod_rem.remainder, lanes);
+            if (!(base_mod + rem_mod)) return AlignCheck::Aligned;
+            else {
+                if (lanes_off) *lanes_off = base_mod + rem_mod;
+                return AlignCheck::Unaligned;
+            }
+        }
+    }
+    void visit(const Load *op) {
+        debug(4) << "HexagonAlignLoads: Working on " << (Expr) op << "..\n";
+        Expr index = mutate(op->index);
+        if (op->type.is_vector()) {
+            bool external = op->image.defined();
+            if (external) {
+                debug(4) << "HexagonAlignLoads: Not dealing with an external image\n";
+                debug(4) << (Expr) op << "\n";
+                expr = op;
+                return;
+            }
+            else {
+                const Ramp *ramp = index.as<Ramp>();
+                const IntImm *stride = ramp ? ramp->stride.as<IntImm>() : NULL;
+                // We will work only on natural vectors supported by the target.
+                if (ramp->lanes != natural_vector_lanes(op->type)) {
+                    expr = op;
+                    return;
+                }
+                if (ramp && stride && stride->value == 1) {
+                    int lanes = ramp->lanes;
+                    // If this is a parameter, the base_alignment should be host_alignment.
+                    // This cannot be an external image because we have already checked for
+                    // it. Otherwise, this is an internal buffers that is always aligned
+                    // to the natural vector width.
+                    int base_alignment = op->param.defined() ?
+                        op->param.host_alignment() : vector_size;
+                    int lanes_off;
+                    AlignCheck ac = get_alignment_check(ramp, base_alignment, &lanes_off);
+                    if (ac == AlignCheck::NoResult) {
+                        // We know nothing about alignment. Give up.
+                        // TODO: Fix this.
+                        debug(4) << "HexagonAlignLoads: Cannot reason about alignment.\n";
+                        debug(4) << "HexagonAlignLoads: Type: " << op->type << "\n";
+                        debug(4) << "HexagonAlignLoads: Index: " << index << "\n";
+                        expr = op;
+                        return;
+                    }
+                    if (ac == AlignCheck::Aligned) {
+                        expr = op;
+                        debug(4) << "HexagonAlignLoads: Encountered a perfectly aligned load.\n";
+                        debug(4) << "HexagonAlignLoads: Type: " << op->type << "\n";
+                        debug(4) << "HexagonAlignLoads: Index: " << index << "\n";
+                        return;
+                    } else {
+                        Expr base = ramp->base;
+                        const Add *add = base.as<Add>();
+                        const IntImm *b = add->b.as<IntImm>();
```

Comment:
Actually, I don't think we need this. We have already run modulus_remainder on ramp->base when we called get_alignment_check above.